cachito/workers/pkg_managers/pip.py

```diff
@@ -0,0 +1,321 @@
+
+    Aims to match setuptools behaviour as closely as possible, but does make
+    some compromises (such as never executing arbitrary Python code).
+    """
+
+    # Valid Python name - any sequence of \w characters that does not start with a number
+    _name_re = re.compile(r"[^\W\d]\w*")
+
+    def __init__(self, top_dir):
+        """
+        Initialize a SetupCFG.
+
+        :param str top_dir: Path to root of project directory
+        """
+        super().__init__(top_dir, "setup.cfg")
+        self._parsed = None
+
+    def get_name(self):
+        """
+        Get metadata.name if present.
+
+        :rtype: str or None
+        """
+        self._parse()
+        name = self._get_option("metadata", "name")
+        if name is None:
+            log.debug("No metadata.name in setup.cfg")
+            return None
+
+        log.debug("Found metadata.name in setup.cfg: %r", name)
+        return name
+
+    def get_version(self):
+        """
+        Get metadata.version if present.
+
+        Partially supports the file: directive (setuptools supports multiple files
+        as an argument to file:, this makes no sense for version).
+
+        Partially supports the attr: directive (will only work if the attribute
+        being referenced is assigned to a literal expression).
+
+        :rtype: str or None
+        """
+        self._parse()
+        version = self._get_option("metadata", "version")
+        if version is None:
+            log.debug("No metadata.version in setup.cfg")
+            return None
+
+        log.debug("Resolving metadata.version in setup.cfg from %r", version)
+        version = self._resolve_version(version)
+        if version is None:
+            log.debug("Failed to resolve metadata.version in setup.cfg")
+            return None
+
+        version = any_to_version(version)
+        log.debug("Found metadata.version in setup.cfg: %r", version)
+        return version
+
+    def _parse(self):
+        """Parse config file if not already parsed."""
+        if self._parsed is None:
+            log.debug("Parsing setup.cfg at %s", self._path)
+            parsed = configparser.ConfigParser()
+            with self._path.open() as f:
+                parsed.read_file(f)
+            self._parsed = parsed
+        return self._parsed
+
+    def _get_option(self, section, option):
+        """Get option from config section, return None if missing."""
+        try:
+            return self._parsed.get(section, option)
+        except (configparser.NoSectionError, configparser.NoOptionError):
+            return None
+
+    def _resolve_version(self, version):
+        """Attempt to resolve the version attribute."""
+        if version.startswith("file:"):
+            file_arg = version[len("file:") :].strip()
+            version = self._read_version_from_file(file_arg)
+        elif version.startswith("attr:"):
+            attr_arg = version[len("attr:") :].strip()
```

Comment:
This is the CI failure (`black` and `flake8` don't agree on the space before `:`) - permission to silence flake8?

Reply:
why space before `:` ?

Reply:
IDK, that's how `black` wants to format it

Reply:
ok, I disagree with black

Reply:
Silenced `black` instead, if anyone disagrees let me know :-)

Reply:
Can we adjust the configuration instead?

Reply:
`black` doesn't have any configuration, from the [README](https://github.com/psf/black/blob/master/README.md):

> Black is a PEP 8 compliant opinionated formatter. Black reformats entire files in place. It is not configurable. It doesn't take previous formatting into account. Your main option of configuring Black is that it doesn't reformat blocks that start with # fmt: off and end with # fmt: on.

We could adjust the `flake8` config, not sure what would be the preference

Reply:
The rule of thumb on this project so far has been to always agree with `black`. Even though we may disagree with some of its choices, it's still worth for all the other parts that it gets correct. I suggest we disable the `E203` check in flake8.

Reply:
Ok, I'll go with `black`. E203 seems valid outside of this specific scenario, I'll just disable the `flake8` check inline

Reply:
Nevermind, `black` will catch the valid cases, disabling in config